File: ChatWithFiles.Domain/Entities/FileAttachment.cs
using ChatWithFiles.Domain.Common;
using ChatWithFiles.Domain.Enums;

namespace ChatWithFiles.Domain.Entities;

public class FileAttachment : BaseEntity
{
    public Guid MessageId { get; set; }
    
    // File info
    public string FileName { get; set; } = string.Empty;
    public string OriginalFileName { get; set; } = string.Empty;
    public string FileType { get; set; } = string.Empty; // MIME type
    public string FileExtension { get; set; } = string.Empty;
    public long FileSize { get; set; }
    
    // Storage
    public string StorageKey { get; set; } = string.Empty;
    public StorageProvider StorageProvider { get; set; } = StorageProvider.AzureBlob;
    public string? StorageRegion { get; set; }
    
    // Upload tracking
    public UploadStatus UploadStatus { get; set; } = UploadStatus.Pending;
    public int UploadProgress { get; set; }
    public DateTime? UploadStartedAt { get; set; }
    public DateTime? UploadCompletedAt { get; set; }
    
    // Post-processing
    public ProcessingStatus ProcessingStatus { get; set; } = ProcessingStatus.Pending;
    public string? ThumbnailUrl { get; set; }
    public bool IsScanned { get; set; }
    public ScanStatus? ScanStatus { get; set; }
    public string? ScanResult { get; set; }
    public DateTime? ScannedAt { get; set; }
    
    // File metadata
    public int? Width { get; set; }
    public int? Height { get; set; }
    public int? Duration { get; set; } // seconds
    public int? BitRate { get; set; }
    public double? FrameRate { get; set; }
    
    // Error handling
    public string? ErrorMessage { get; set; }
    public string? ErrorCode { get; set; }
    public int RetryCount { get; set; }
    public int MaxRetries { get; set; } = 3;
    
    // Audit
    public Guid UploadedBy { get; set; }
    
    // Navigation properties
    public ChatMessage? Message { get; set; }
    public User? Uploader { get; set; }
    public ICollection<UploadSession> UploadSessions { get; set; } = new List<UploadSession>();
    
    // Helper methods
    public void UpdateProgress(int progress, UploadStatus status)
    {
        UploadProgress = progress;
        UploadStatus = status;
        UpdatedAt = DateTime.UtcNow;
        
        if (status == UploadStatus.Uploading && UploadStartedAt == null)
        {
            UploadStartedAt = DateTime.UtcNow;
        }
    }
    
    public void MarkAsUploaded(string storageKey, long fileSize)
    {
        UploadStatus = UploadStatus.Uploaded;
        UploadProgress = 100;
        UploadCompletedAt = DateTime.UtcNow;
        ProcessingStatus = ProcessingStatus.Pending;
        StorageKey = storageKey;
        FileSize = fileSize;
        UpdatedAt = DateTime.UtcNow;
    }
    
    public void MarkAsProcessing()
    {
        ProcessingStatus = ProcessingStatus.InProgress;
        UpdatedAt = DateTime.UtcNow;
    }
    
    public void MarkProcessingCompleted(string? thumbnailUrl = null)
    {
        ProcessingStatus = ProcessingStatus.Completed;
        ThumbnailUrl = thumbnailUrl;
        UpdatedAt = DateTime.UtcNow;
    }
    
    public void MarkProcessingFailed(string error, string? errorCode = null)
    {
        ProcessingStatus = ProcessingStatus.Failed;
        ErrorMessage = error;
        ErrorCode = errorCode;
        RetryCount++;
        UpdatedAt = DateTime.UtcNow;
    }
    
    public void MarkScanned(ScanStatus status, string? result = null)
    {
        IsScanned = true;
        ScanStatus = status;
        ScanResult = result;
        ScannedAt = DateTime.UtcNow;
        UpdatedAt = DateTime.UtcNow;
    }
    
    public bool CanRetry() => RetryCount < MaxRetries;
    
    public void PrepareForRetry()
    {
        UploadStatus = UploadStatus.Pending;
        UploadProgress = 0;
        ProcessingStatus = ProcessingStatus.Pending;
        UpdatedAt = DateTime.UtcNow;
    }
}

File: FlutterApp/lib/domain/entities/file_attachment.dart
import 'package:equatable/equatable.dart';

enum UploadStatus {
  pending,
  uploading,
  uploaded,
  processing,
  completed,
  failed,
  cancelled
}

enum ProcessingStatus {
  pending,
  inProgress,
  completed,
  failed,
  skipped
}

class FileAttachment extends Equatable {
  final String id;
  final String messageId;
  final String fileName;
  final String originalFileName;
  final String fileType;
  final String fileExtension;
  final int fileSize;
  final String storageKey;
  final UploadStatus uploadStatus;
  final int uploadProgress;
  final ProcessingStatus processingStatus;
  final String? thumbnailUrl;
  final bool isScanned;
  final String? scanStatus;
  final int? width;
  final int? height;
  final int? duration;
  final String? errorMessage;
  final int retryCount;
  final DateTime createdAt;
  final DateTime? updatedAt;
  final String? localFilePath; // For offline support
  
  const FileAttachment({
    required this.id,
    required this.messageId,
    required this.fileName,
    required this.originalFileName,
    required this.fileType,
    required this.fileExtension,
    required this.fileSize,
    this.storageKey = '',
    this.uploadStatus = UploadStatus.pending,
    this.uploadProgress = 0,
    this.processingStatus = ProcessingStatus.pending,
    this.thumbnailUrl,
    this.isScanned = false,
    this.scanStatus,
    this.width,
    this.height,
    this.duration,
    this.errorMessage,
    this.retryCount = 0,
    required this.createdAt,
    this.updatedAt,
    this.localFilePath,
  });
  
  FileAttachment copyWith({
    UploadStatus? uploadStatus,
    int? uploadProgress,
    ProcessingStatus? processingStatus,
    String? thumbnailUrl,
    String? errorMessage,
    int? retryCount,
    String? storageKey,
    String? localFilePath,
  }) {
    return FileAttachment(
      id: id,
      messageId: messageId,
      fileName: fileName,
      originalFileName: originalFileName,
      fileType: fileType,
      fileExtension: fileExtension,
      fileSize: fileSize,
      storageKey: storageKey ?? this.storageKey,
      uploadStatus: uploadStatus ?? this.uploadStatus,
      uploadProgress: uploadProgress ?? this.uploadProgress,
      processingStatus: processingStatus ?? this.processingStatus,
      thumbnailUrl: thumbnailUrl ?? this.thumbnailUrl,
      isScanned: isScanned,
      scanStatus: scanStatus,
      width: width,
      height: height,
      duration: duration,
      errorMessage: errorMessage ?? this.errorMessage,
      retryCount: retryCount ?? this.retryCount,
      createdAt: createdAt,
      updatedAt: DateTime.now(),
      localFilePath: localFilePath ?? this.localFilePath,
    );
  }
  
  bool get isImage => fileType.startsWith('image/');
  bool get isVideo => fileType.startsWith('video/');
  bool get isDocument => fileType.startsWith('application/') || 
                         fileType.startsWith('text/');
  bool get canRetry => retryCount < 3 && uploadStatus == UploadStatus.failed;
  
  String get formattedFileSize {
    if (fileSize < 1024) return '$fileSize B';
    if (fileSize < 1024 * 1024) return '${(fileSize / 1024).toStringAsFixed(1)} KB';
    return '${(fileSize / (1024 * 1024)).toStringAsFixed(1)} MB';
  }
  
  @override
  List<Object?> get props => [id, messageId, fileName, uploadStatus, uploadProgress];
}
